lib/board.rb

```diff
@@ -45,41 +42,89 @@ def place(ship, ship_coordinates)
     else false
     end
   end
-
+  
   def render(view = false)
     if view == true
       rendered_true
     else
       rendered
     end
   end
-
+  
   #helper methods
   
+  def valid_coordinate?(coordinate)
+    @cells.has_key?(coordinate)
+  end
+  
   def rendered_true
     "  1 2 3 4 \n" +
     "A #{@cells["A1"].render(true)} #{@cells["A2"].render(true)} #{@cells["A3"].render(true)} #{@cells["A4"].render(true)}\n" +
     "B #{@cells["B1"].render(true)} #{@cells["B2"].render(true)} #{@cells["B3"].render(true)} #{@cells["B4"].render(true)}\n" +
     "C #{@cells["C1"].render(true)} #{@cells["C2"].render(true)} #{@cells["C3"].render(true)} #{@cells["C4"].render(true)}\n" +
     "D #{@cells["D1"].render(true)} #{@cells["D2"].render(true)} #{@cells["D3"].render(true)} #{@cells["D4"].render(true)}"
   end
-
+  
   def rendered
     "  1 2 3 4 \n" +
     "A #{@cells["A1"].render} #{@cells["A2"].render} #{@cells["A3"].render} #{@cells["A4"].render}\n" +
     "B #{@cells["B1"].render} #{@cells["B2"].render} #{@cells["B3"].render} #{@cells["B4"].render}\n" +
     "C #{@cells["C1"].render} #{@cells["C2"].render} #{@cells["C3"].render} #{@cells["C4"].render}\n" +
     "D #{@cells["D1"].render} #{@cells["D2"].render} #{@cells["D3"].render} #{@cells["D4"].render}"
   end
-
+  
   def validate_cell(ship_coordinates)
-    ship_coordinates.all? do |cell|
+    ship_coordinates.each.all? do |cell|
       cells[cell].empty?
     end
   end
 
+  def coordinates_consecutive_horizontal?(coordinates)
+    if @horizontal_cells[0].each_cons(coordinates.length).any? do |cells|
+       coordinates == cells
+      end
+        true
+      elsif @horizontal_cells[1].each_cons(coordinates.length).any? do |cells|
```

Comment:
I love that we finally incorporated each_cons into the code. It seems to be doing the trick for player.